### PR TITLE
Adjustments for future LLVM-18 release

### PR DESCRIPTION
**Pull request by linehill**
Updates are currently based on:
* llvm-project: b12a742c20ae548c5d84f1719f9199577b382ccd
* llvm-spirv: 4e9edf69b836eaf0540df9c43ee95d908b2f1f77

### CODE REVIEW

**Comment by franz** [comment]
Tested with LLVM main branch (probably not exactly same commits, built on Nov 28), chipStar & catch tests build OK, check.py tests pass with iGPU (both L0 and OCL backends).

**Review by pjaaskel — APPROVED** [review]
LLVM 18.1 has been branched, so I suggest to pull this in. Do all the tests pass?

**Comment by pvelesko** [comment]
@linehill no longer a draft perhaps?

**Comment by linehill** [comment]
I'll test with the recent revision of the LLVM-18 locally, update if needed and mark then this for review.

**Comment by pvelesko** [comment]
Is this using mainline LLVM and SPIRV-Translator? I didn't find new branches in our fork for LLVM18.

Using mainline seems like tests pass locally. I updated the llvm_configure.sh script and will update CI to add 18 to testing. 

I had to add `-DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD="SPIRV"` option because without it I got a lot of warnings from `opt` about `spirv64` target not being built. These warnings went away with the experimental target enabled. Expected? @linehill

File: scripts/unit_tests.sh [hunk 1]
```diff
@@ -37,9 +37,12 @@ elif [ "$2" == "llvm-16" ]; then
 elif [ "$2" == "llvm-17" ]; then
   LLVM=llvm-17
   CLANG=llvm/17.0-unpatched-spirv
+elif [ "$2" == "llvm-18" ]; then
+  LLVM=llvm-18
+  CLANG=llvm/18.0-unpatched-spirv
 else
   echo "$2"
-  echo "Invalid 2nd argument. Use either 'llvm-15', 'llvm-16' or 'llvm-17'."
+  echo "Invalid 2nd argument. Use either 'llvm-15', 'llvm-16', 'llvm-17' ir 'llvm-18'."
```
Comment (pjaaskel) [inline scripts/unit_tests.sh:45]:
```suggestion
  echo "Invalid 2nd argument. Use either 'llvm-15', 'llvm-16', 'llvm-17' or 'llvm-18'."
```

**Pull request merged**